textile/chat-features.textile

```diff
@@ -9,5 +9,148 @@ jump_to:
+
+h4(#rooms-lifecycle-operations). Room Lifecycle Operations
+
+* @(CHA-RL1)@ A room must be explicitly @ATTACHED@ to.
+** @(CHA-RL1a)@ @[Testable]@ If the room is already in the @ATTACHED@ status, then this operation is no-op.
+** @(CHA-RL1b)@ @[Testable]@ If the room is in the @RELEASING@ status, the operation shall be rejected with error code @102102@.
+** @(CHA-RL1c)@ @[Testable]@ If the room is in the @RELEASED@ status, the operation shall be rejected with error code @102103@.
+** @(CHA-RL1d)@ @[Testable]@ If a room lifecycle operation is already in progress, this operation shall wait until the current operation completes before continuing.
+** @(CHA-RL1e)@ @[Testable]@ Notwithstanding the above points, when the attachment operation begins, the room shall be transitioned to the @ATTACHING@ status.
+** @(CHA-RL1f)@ @[Testable]@ The underlying @contributors@ will have their Realtime Channels attached in sequence.
+** @(CHA-RL1g)@ When all @contributors@ Realtime Channels successfully attach, the operation is now complete and the room is considered attached.
+*** @(CHA-RL1g1)@ @[Testable]@ The room status shall be transitioned to @ATTACHED@.
+*** @(CHA-RL1g2)@ @[Testable]@ Any contributors to room that have a pending discontinuity event against them, must be notified of this fact, using the error that caused the original discontinuity.
+*** @(CHA-RL1g3)@ @[Testable]@ Any transient disconnect timeouts shall be cleared.
+** @(CHA-RL1h)@ If a one of the @contributors@ Realtime Channels fails to attach, then the operation has failed and must be rolled back.
+*** @(CHA-RL1h1)@ @[Testable]@ The @attach@ call must throw an @ErrorInfo@. The code for this error is determined by the feature that failed (see the @ErrorCodes@ enum in the JS SDK for more information).
+*** @(CHA-RL1h2)@ @[Testable]@ If the underlying Realtime Channel entered the @SUSPENDED@ state, then the room status will transition to @SUSPENDED@, using the Realtime Channel error as the @cause@ field of the @ErrorInfo@.
+*** @(CHA-RL1h3)@ @[Testable]@ When the room enters the @SUSPENDED@ status, the room lifecycle shall enter the recovery loop (TODO: REF) and attempt to re-establish attachment. The attach operation shall fail using the @ErrorInfo@ associated with the @SUSPENDED@ room status as the error.
+*** @(CHA-RL1h4)@ @[Testable]@ If the underlying Realtime Channel entered the @FAILED@ state, then the room status will transition to @FAILED@, using the Realtime Channel error as the @cause@ field of the @ErrorInfo@.
+*** @(CHA-RL1h5)@ @[Testable]@ When the room enters the @FAILED@ status, then the room will detached all channels that are not in the @FAILED@ state.
+*** @(CHA-RL1h6)@ @[Testable]@ When the room enters the @FAILED@ status, and a channel fails to detach on command, then the detach operation will be repeated until such a time that all channels have detached successfully.
```

Comment:
When in the "a channel entered FAILED" state, it's not clear how the room `ATTACH` operation should end — I’m guessing that it waits for the channels to have all detached, and then it completes (with an error?)

Reply:
Clarified!

Reply:
Clarified!